network/p2p/node_sampler_test.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	safemath "github.com/ava-labs/avalanchego/utils/math"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+func TestPeersSample(t *testing.T) {
+	nodeID1 := ids.GenerateTestNodeID()
+	nodeID2 := ids.GenerateTestNodeID()
+	nodeID3 := ids.GenerateTestNodeID()
+
+	tests := []struct {
+		name         string
+		connected    set.Set[ids.NodeID]
+		disconnected set.Set[ids.NodeID]
+		n            int
+	}{
+		{
+			name: "no peers",
+			n:    1,
+		},
+		{
+			name:      "one peer connected",
+			connected: set.Of(nodeID1),
+			n:         1,
+		},
+		{
+			name:      "multiple peers connected",
+			connected: set.Of(nodeID1, nodeID2, nodeID3),
+			n:         1,
+		},
+		{
+			name:         "peer connects and disconnects - 1",
+			connected:    set.Of(nodeID1),
+			disconnected: set.Of(nodeID1),
+			n:            1,
+		},
+		{
+			name:         "peer connects and disconnects - 2",
+			connected:    set.Of(nodeID1, nodeID2),
+			disconnected: set.Of(nodeID2),
+			n:            1,
+		},
+		{
+			name:         "peer connects and disconnects - 2",
+			connected:    set.Of(nodeID1, nodeID2, nodeID3),
+			disconnected: set.Of(nodeID1, nodeID2),
+			n:            1,
+		},
+		{
+			name:      "less than n peers",
+			connected: set.Of(nodeID1, nodeID2, nodeID3),
+			n:         4,
+		},
+		{
+			name:      "n peers",
+			connected: set.Of(nodeID1, nodeID2, nodeID3),
+			n:         3,
+		},
+		{
+			name:      "more than n peers",
+			connected: set.Of(nodeID1, nodeID2, nodeID3),
+			n:         2,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require := require.New(t)
+			sampler := &Peers{}
+
+			for connected := range tt.connected {
+				require.NoError(sampler.Connected(context.Background(), connected, nil))
+			}
+
+			for disconnected := range tt.disconnected {
+				require.NoError(sampler.Disconnected(context.Background(), disconnected))
```

Comment:
why not use context.WithCancel for the test?

Reply:
Hum can you elaborate? I didn't understand why this would be helpful here.